module/trie.c

```diff
@@ -0,0 +1,286 @@
+/* This file implements a new module native data type called "TRIE".
+ * The data structure is based on the idea of a trie (aka a prefix 
+ * tree) and largely works the same way.
+ */
+#define IN_TRIE 1
+#define NOT_IN_TRIE 0 
+#define PARTIAL_IN_TRIE (-1)
+
+#include "../redismodule.h"
+#include <stdlib.h>
+#include <string.h>
+#include <assert.h>
+#include <math.h>
+
+
+static RedisModuleType *trie;
+
+/* ========================== Internal data structure (Bare bones functions)  ======================= */
+
+struct trie {
+    char current; 
+        // The first trie_t will be '/0' for any Trie.
+    struct trie **children;
+         // ALPHABET_SIZE is 256 for all possible characters.
+    int is_word; 
+        // if is_word is 1, indicates that this is the end of a word. Otherwise 0.
+    struct trie *parent;
+        // parent trie for traversing backwards
+};
+
+/*
+    Creates and allocates memory for new trie.
+    
+    Parameters:
+     - current: A char for the current character
+    
+    Returns:
+     - A pointer to the trie, or REDISMODULE_ERR if a pointer 
+       cannot be allocated
+*/
+struct trie *new_trie (char current)
+{
+    struct trie *t = RedisModule_Calloc(1,sizeof(*t));
+    
+    if (t == NULL){
+        fprintf(stderr, "Could not allocate memory for trie\n");
+        return NULL;
+    } 
+
+    t->current = current;
+
+    t->children = RedisModule_Calloc(256,sizeof(t));
+    if (t->children == NULL){
+        fprintf(stderr, "Could not allocate memory for t->children\n");
+        return NULL;
+    }
+
+    t->is_word = 0;
+    t->parent = NULL;
+    return t;
+}
+
+/*
+    Free an entire trie.
+    Parameters:
+     - t: A trie pointer
+    
+    Returns:
+     - Always returns 0
+*/
+int trie_free(struct trie *t)
+{
+    if (t != NULL) {
+
+        for (int i=0; i<256; i++ ){
+            if (t->children[i] !=NULL)
+                trie_free(t->children[i]);
+        }
+
+        RedisModule_Free(t);
+    }
+    return 0;
+}
+
+/*
+    Creates new node in trie.
+    Parameters:
+     - current: A char indicating the character of the node being added
+     - t: A pointer to the trie where the node is to be added
+    
+    Returns:
+     - 0 on success, 1 if an error occurs.
+    Details: 
+     - Set t->children[current] to be current
+     - is_word for new node set to 0.
+*/
+int add_node(char current, struct trie *t)
```

Comment:
All trie functions should start with `trie` and should have the `struct trie *` as the first argument. For example, this function should be `trie_add_node(struct trie *t, char current)`. Please make sure all functions adhere to this convention.

Reply:
@dengyijia @valeriehan6 could you guys change this in the implementation as well?